File: python/src/codex_harness_kit/connection.py
from __future__ import annotations

from collections.abc import AsyncIterator

from .codec import JsonRpcCodec
from .errors import JsonRpcCodecError, ProtocolStreamError
from .messages import JsonRpcMessage
from .transport import Transport


class ProtocolConnection:
    def __init__(self, transport: Transport) -> None:
        self._transport = transport

    async def send(self, message: JsonRpcMessage) -> None:
        await self._transport.send(JsonRpcCodec.encode(message))

    def __aiter__(self) -> AsyncIterator[JsonRpcMessage]:
        return self._iterate()

    async def close(self) -> None:
        await self._transport.close()

    async def _iterate(self) -> AsyncIterator[JsonRpcMessage]:
        async for frame in self._transport:
            try:
                yield JsonRpcCodec.decode(frame)
            except JsonRpcCodecError as exc:
                raise ProtocolStreamError(
                    f"Failed to decode incoming JSON-RPC frame: {exc}"
                ) from exc

File: typescript/src/connection.ts
import { JsonRpcCodec } from "./codec.js";
import { JsonRpcCodecError, ProtocolStreamError } from "./errors.js";
import type { JsonRpcMessage } from "./messages.js";
import type { Transport } from "./transport.js";

export class ProtocolConnection implements AsyncIterable<JsonRpcMessage> {
  constructor(private readonly transport: Transport) {}

  async send(message: JsonRpcMessage): Promise<void> {
    await this.transport.send(JsonRpcCodec.encode(message));
  }

  async close(): Promise<void> {
    await this.transport.close();
  }

  async *[Symbol.asyncIterator](): AsyncIterator<JsonRpcMessage> {
    for await (const frame of this.transport) {
      try {
        yield JsonRpcCodec.decode(frame);
      } catch (error: unknown) {
        if (error instanceof JsonRpcCodecError) {
          throw new ProtocolStreamError(
            `Failed to decode incoming JSON-RPC frame: ${error.message}`,
            { cause: error },
          );
        }
        throw error;
      }
    }
  }
}
